File: asr/android/sdk/src/main/java/com/amphion/asr/internal/StreamingAgcIngress.kt
package com.amphion.asr.internal

import java.io.Closeable

/** Produces paired raw/processed frames for the session in submission order. */
internal interface AgcFrameProcessor : Closeable {
    fun process(samples: FloatArray): List<ProcessedAudioFrame>
    fun flush(): List<ProcessedAudioFrame>
}

/**
 * Owns the session-to-AGC seam: every produced frame is delivered exactly once and a flush drains
 * the processor remainder through the same consumer before the session changes stream state.
 */
internal class StreamingAgcIngress(
    private val processor: AgcFrameProcessor,
) : Closeable {
    private var closed = false

    fun accept(samples: FloatArray, consume: (ProcessedAudioFrame) -> Unit) {
        check(!closed) { "AGC ingress is closed" }
        processor.process(samples).forEach(consume)
    }

    fun flush(consume: (ProcessedAudioFrame) -> Unit) {
        check(!closed) { "AGC ingress is closed" }
        processor.flush().forEach(consume)
    }

    override fun close() {
        if (closed) return
        closed = true
        processor.close()
    }
}

File: asr/harmony/sdk/src/main/ets/com/amphion/asr/StreamingAgcIngress.ts
import { ProcessedAudioFrame } from './StreamingAgcProcessor';

export interface AgcFrameProcessor {
  process(samples: Float32Array): ProcessedAudioFrame[];
  flush(): ProcessedAudioFrame[];
  close(): void;
}

export type AgcFrameConsumer = (frame: ProcessedAudioFrame) => void;
export type AsyncAgcFrameConsumer = (frame: ProcessedAudioFrame) => Promise<void>;

/** Delivers every paired AGC frame once, in order, through both sync and async session lanes. */
export class StreamingAgcIngress {
  private processor: AgcFrameProcessor;
  private closed: boolean = false;

  constructor(processor: AgcFrameProcessor) {
    this.processor = processor;
  }

  accept(samples: Float32Array, consume: AgcFrameConsumer): void {
    if (this.closed) throw new Error('AGC ingress is closed');
    const frames = this.processor.process(samples);
    for (let index = 0; index < frames.length; index++) consume(frames[index]);
  }

  async acceptAsync(samples: Float32Array, consume: AsyncAgcFrameConsumer): Promise<void> {
    if (this.closed) throw new Error('AGC ingress is closed');
    const frames = this.processor.process(samples);
    for (let index = 0; index < frames.length; index++) await consume(frames[index]);
  }

  flush(consume: AgcFrameConsumer): void {
    if (this.closed) throw new Error('AGC ingress is closed');
    const frames = this.processor.flush();
    for (let index = 0; index < frames.length; index++) consume(frames[index]);
  }

  async flushAsync(consume: AsyncAgcFrameConsumer): Promise<void> {
    if (this.closed) throw new Error('AGC ingress is closed');
    const frames = this.processor.flush();
    for (let index = 0; index < frames.length; index++) await consume(frames[index]);
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.processor.close();
  }
}
